### DISCUSSION TITLE
How to better contribute exceljs ?

### DISCUSSION THREAD

**Discussion by skypesky** [exceljs/exceljs#1683] [Q&A] [answered]
In order to understand the working principle and better contribution of `exceljs`, besides the source code, what standard documents do I need to see. Can you provide some links to related documents? At present, I only know that. Xlsx files are basically composed of some XML, but I don't know why workbooks should have worksheets attribute and so on.

@Siemienik @guyonroche @alubbe   
You are experts in this field. Can you give me some suggestions?(I'm really sorry to disturb you!)
Thank you very much.

**Comment by Siemienik** [comment, marked as answer]
Hi @skypesky, great that you are asking :)

### XLSX Format

XLSX is a standardized format: ECMA 376 (https://www.ecma-international.org/publications-and-standards/standards/ecma-376/)
I've forked XML schema for that standard: https://github.com/exceljs/ooxml-xsd

For instance, if you want to know members of the workbook (in the XLSX format), you may search `workbook` and finally go to  https://github.com/exceljs/ooxml-xsd/blob/master/sml-workbook.xsd . There are listed all workbook properties and their types.

1. Workbook.sheet prop:

https://github.com/exceljs/ooxml-xsd/blob/aba7f40dc52d9a3cfe236e9e3e15a5148d5ca999/sml-workbook.xsd#L37-L41

Type is `CT_Sheets`

2. Following this:

https://github.com/exceljs/ooxml-xsd/blob/aba7f40dc52d9a3cfe236e9e3e15a5148d5ca999/sml-workbook.xsd#L415-L423

`CT_Sheets` is an array of `CT_Sheet`

3. So let's look what is CT_Sheet:

https://github.com/exceljs/ooxml-xsd/blob/aba7f40dc52d9a3cfe236e9e3e15a5148d5ca999/sml-workbook.xsd#L424-L445

Where there most important is `r:id` point to sheet file, described in:

https://github.com/exceljs/ooxml-xsd/blob/aba7f40dc52d9a3cfe236e9e3e15a5148d5ca999/sml-sheet.xsd

etc. etc.

### Reading and writing files

Xlsx is actually a ZIP file, so it is unzipped, and then We are using `saxes` packages to read and writing files.

The logic for handling xlsx file is kept in https://github.com/exceljs/exceljs/tree/master/lib/xlsx/xform

### Examining content of XLSX file:

As the XLSX is a ZIP file, it is easy to examine what it contains:

```bash
unzip -d dist what-is-inside-this-file.xlsx
cd dist
```

I open this folder in a WebStorm and format files.

It could be even modified and packed reverse into XLSX file by a commend:

```bash
cd dist
find . -type f | xargs zip ../my-modified-xlsx.xlsx
cd ../
```

**Reply by skypesky** [reply]
@Siemienik 
   Thank you!
   Your help is very important to me. Now I have a clear direction. I believe that through my research, I can have a deeper understanding of the principle of XLSX(although I may encounter difficulties on the way)!

**Reply by Siemienik** [reply]
If any difficulties have you,  just ask me on gitter: https://gitter.im/Siemienik/community :)
